File: hw5/assignment/SplitFileClient.go
/*
SplitFileClient.go
20190532 sang yun lee
*/
package main

import (
	"bufio"
	"fmt"
	"io"
	"net"
	"os"
	"path/filepath"
	"strconv"
	"sync"
)

func main() {
	// argument handling
	if len(os.Args) != 3 {
		fmt.Println("Invalid argument.")
		os.Exit(1)
	}
	// get argument from system args
	commandName := os.Args[1]
	fileName := os.Args[2]

	// server Info hardcoding
	firstServerName := "127.0.0.1"
	firstServerPort := "40532"
	secondServerName := "127.0.0.1"
	secondServerPort := "50532"

	var wg sync.WaitGroup

	// put command case
	if commandName == "put" {
		wg.Add(2)
		// sendFile odd byte
		go sendFile(fileName, firstServerName, firstServerPort, 0, &wg)
		// sendFile even byte
		go sendFile(fileName, secondServerName, secondServerPort, 1, &wg)
		wg.Wait()
	} else if os.Args[1] == "get" { // get command case
		wg.Add(2)
		// receiveFile odd byte
		go receiveFile(fileName, firstServerName, firstServerPort, 0, &wg)
		// receiveFile even byte
		go receiveFile(fileName, secondServerName, secondServerPort, 1, &wg)
		wg.Wait()
		// get file extension from fileName
		fileExtension := filepath.Ext(fileName)
		// get file name without extension from fileName
		fileName = fileName[0 : len(fileName)-len(fileExtension)]
		// merged file create
		file, err := os.Create(fileName + "-merged" + fileExtension)
		if err != nil {
			fmt.Println("fail to create file:", err)
			os.Exit(1)
		}
		// set Temp file name
		tmpFileName1 := fmt.Sprintf("%s-part%d%stmp%s", fileName, 1, fileExtension, fileExtension)
		tmpFileName2 := fmt.Sprintf("%s-part%d%stmp%s", fileName, 2, fileExtension, fileExtension)
		// open Temp file
		tmpFile1, _ := os.Open(tmpFileName1)
		tmpFile2, _ := os.Open(tmpFileName2)
		// set close Function and delete tmp file
		defer func(tmpFile1 *os.File) {
			err := tmpFile1.Close()
			if err != nil {
				fmt.Println("File close error")
				os.Exit(1)
			}
			err = os.Remove(tmpFileName1)
			if err != nil {
				fmt.Println("File Remove error")
				os.Exit(1)
			}
		}(tmpFile1)
		defer func(tmpFile2 *os.File) {
			err := tmpFile2.Close()
			if err != nil {
				fmt.Println("File close error")
			}
			err = os.Remove(tmpFileName2)
			if err != nil {
				fmt.Println("File Remove error")
			}
		}(tmpFile2)

		// set Buffer from each tmp file
		reader1 := bufio.NewReader(tmpFile1)
		reader2 := bufio.NewReader(tmpFile2)
		// check even, odd
		var byteCnt int64
		var finishCnt int64
		for {
			// all file is finished, exit
			if finishCnt == 2 {
				break
			}
			if byteCnt%2 == 0 {
				// get 1 byte from odd file
				b, err := reader1.ReadByte()
				if err != nil {
					finishCnt++
					continue // finish to reach file end
				}
				_, err = file.Write([]byte{b})
				if err != nil {
					return
				}
			} else {
				// get 1 byte from even file
				b, err := reader2.ReadByte()
				if err != nil {
					finishCnt++
					continue // finish to reach file end
				}
				_, err = file.Write([]byte{b})
				if err != nil {
					return
				}
			}
			byteCnt++
		}

		fmt.Printf("%s%s file merge sucessful!\n", fileName, fileExtension)
	} else { // otherwise case
		fmt.Println("Invalid argument.")
	}
}

// put file to server
func sendFile(fileName string, serverName string, serverPort string, part int, wg *sync.WaitGroup) {
	defer wg.Done()
	// connect to TCP server
	conn, err := net.Dial("tcp", serverName+":"+serverPort)
	if err != nil {
		fmt.Println("fail to connect server: ", err)
		os.Exit(1)
	}
	defer func(conn net.Conn) {
		err := conn.Close()
		if err != nil {
			fmt.Println("socket close error : ", err)
			os.Exit(1)
		}
	}(conn)

	// prepare command string to send command
	commandStr := "put"
	_, err = conn.Write([]byte(commandStr))
	if err != nil {
		fmt.Println("Write failed")
		os.Exit(1)
	}
	commandBuffer := make([]byte, 1024)
	read, _ := conn.Read(commandBuffer)
	// is server response is not ok
	if string(commandBuffer[:read]) != "ok" {
		fmt.Println("fail to receive fileName")
		os.Exit(1)
	}
	fmt.Println("Request to server to put :" + fileName)

	// try to open file
	originalFile, err := os.Open(fileName)
	if err != nil {
		fmt.Println("File Open failed")
		os.Exit(1)
	}
	defer func(originalFile *os.File) {
		err := originalFile.Close()
		if err != nil {
			fmt.Println("File Closed failed")
			os.Exit(1)
		}
	}(originalFile)

	// get file extension from fileName
	fileExtension := filepath.Ext(fileName)
	// get file name without extension from fileName
	fileName = fileName[0 : len(fileName)-len(fileExtension)]
	// join that strings
	fileName = fmt.Sprintf("%s-part%d%s", fileName, part+1, fileExtension)
	// send to target File name
	_, err = conn.Write([]byte(fileName))
	if err != nil {
		fmt.Println("Write failed")
		os.Exit(1)
	}
	fileNameBuffer := make([]byte, 1024)
	read, _ = conn.Read(fileNameBuffer)
	if string(fileNameBuffer[:read]) != "ok" {
		fmt.Println("fail to receive command")
		os.Exit(1)
	}

	// try to get file Stats
	fileInfo, err := originalFile.Stat()
	if err != nil {
		fmt.Println("fail to get file stat:", err)
		return
	}

	// convert file Size string to Integer
	size := strconv.FormatInt(fileInfo.Size(), 10)
	// send file Size to server
	_, err = conn.Write([]byte(size))
	if err != nil {
		fmt.Println("Write failed")
		os.Exit(1)
	}
	fileSizeBuffer := make([]byte, 1024)
	read, _ = conn.Read(fileSizeBuffer)
	if string(fileSizeBuffer[:read]) != "ok" {
		fmt.Println("fail to receive fileSize")
		os.Exit(1)
	}

	// get Buffer from file
	reader := bufio.NewReader(originalFile)
	var byteCount int
	for {
		// get 1 byte from file
		b, err := reader.ReadByte()
		if err != nil {
			break // finish to reach file end
		}

		// check odd, even byte
		if byteCount%2 == part {
			_, err := conn.Write([]byte{b})
			if err != nil {
				fmt.Println("Write failed")
				os.Exit(1)
			}
		}
		byteCount++
	}
	fmt.Printf("%s send successful\n", fileName)
}

func receiveFile(fileName string, serverName string, serverPort string, part int, wg *sync.WaitGroup) {
	// function finished, notify to wait group
	defer wg.Done()

	fmt.Println("Request to server to get :" + fileName)
	// connect to TCP server
	conn, err := net.Dial("tcp", serverName+":"+serverPort)
	if err != nil {
		fmt.Println("fail to connect server: ", err)
		return
	}

	// prepare command string to send command
	commandStr := "get"
	_, err = conn.Write([]byte(commandStr))
	if err != nil {
		fmt.Println("Write failed")
		os.Exit(1)
	}
	commandBuffer := make([]byte, 1024)
	read, _ := conn.Read(commandBuffer)
	// is server response is not ok
	if string(commandBuffer[:read]) != "ok" {
		fmt.Println(string(commandBuffer[:read]))
		os.Exit(1)
	}

	// get file extension from fileName
	fileExtension := filepath.Ext(fileName)
	// get file name without extension from fileName
	fileName = fileName[0 : len(fileName)-len(fileExtension)]
	// join that strings
	fileName = fmt.Sprintf("%s-part%d%s", fileName, part+1, fileExtension)
	// request file Name
	_, err = conn.Write([]byte(fileName))
	if err != nil {
		fmt.Println("Write failed")
		os.Exit(1)
	}

	// try to get fileSize
	fileSizeBuffer := make([]byte, 1024)
	read, _ = conn.Read(fileSizeBuffer)
	fileSize, err := strconv.ParseInt(string(fileSizeBuffer[:read]), 10, 64)
	if err != nil {
		fmt.Println("fail to transfer file size:", err)
		_, err = conn.Write([]byte("fail to transfer file size"))
		if err != nil {
			fmt.Println("Write failed")
			os.Exit(1)
		}
		os.Exit(1)
	} else {
		_, err = conn.Write([]byte("ok"))
		if err != nil {
			fmt.Println("Write failed")
			os.Exit(1)
		}
	}

	file, err := os.Create(fileName + "tmp" + filepath.Ext(fileName))
	if err != nil {
		fmt.Println("fail to create file:", err)
		return
	}
	// received file contents and saved
	var receivedBytes int64
	buffer := make([]byte, 1024)
	for {
		n, err := conn.Read(buffer)
		if err != nil {
			if err != io.EOF {
				fmt.Println("fail to read file:", err)
			}
			break
		}
		receivedBytes += int64(n)

		// write to file from server contents
		if _, err := file.Write(buffer[:n]); err != nil {
			fmt.Println("fail to write file:", err)
			return
		}

		// check file is read finished
		if receivedBytes >= fileSize {
			break
		}
	}
	fmt.Printf("%s file store sucessful!\n", fileName)
}

File: hw5/rust_bonus_assignment/src/SplitFileClient.rs
/*
SplitFileClient.rs
20190532 sang yun lee
*/

use std::env::args;
use std::io::{Read, Write, BufReader};
use std::net::TcpStream;
use std::path::Path;
use std::process::exit;
use std::fs::File;
use std::fs::remove_file;
use std::thread;

// message buffer size
const MSG_SIZE: usize = 1024;
// server info
const SERVER_IP1: &str = "127.0.0.1";
const SERVER_PORT1: &str = "40532";
const SERVER_IP2: &str = "127.0.0.1";
const SERVER_PORT2: &str = "50532";

fn main() {
	// get system argument
	let args: Vec<String> = args().collect();
	if args.len() != 3 {
		println!("Invalid argument");
		exit(1);
	}
	// get argument from system args
	let command_name = &args[1];
	let file_name = args[2].clone();

	if command_name == "put" {
		let file_name1 = file_name.clone();
		let file_name2 = file_name.clone();
		let handle1 = thread::spawn(move || {
			send_file(&file_name1, SERVER_IP1, SERVER_PORT1, 0);
		});
		let handle2 = thread::spawn(move || {
			send_file(&file_name2, SERVER_IP2, SERVER_PORT2, 1);
		});
		handle1.join().unwrap();
        handle2.join().unwrap();
	} else if command_name == "get" {
		let file_name1 = file_name.clone();
		let file_name2 = file_name.clone();
		let handle1 = thread::spawn(move || {
			receive_file(&file_name1, SERVER_IP1, SERVER_PORT1, 0);
		});
		let handle2 = thread::spawn(move || {
			receive_file(&file_name2, SERVER_IP2, SERVER_PORT2, 1);
		});
		handle1.join().unwrap();
        handle2.join().unwrap();

		// get file extension from fileName
		let file_extension = match Path::new(&file_name).extension().and_then(|s| s.to_str()) {
			Some(ext) => ".".to_string() + ext,
			None => "".to_string()
		};
		let file_name = &file_name[0..(file_name.len() - file_extension.len())];
		// set tmp file name
		let tmp_file_name1 = format!("{}-part{}{}tmp{}", &file_name, 1, file_extension, file_extension);
		let tmp_file_name2 = format!("{}-part{}{}tmp{}", &file_name, 2, file_extension, file_extension);
		// make result file name
		let file_name = format!("{}-merged{}", file_name, file_extension);
		// file open
		let tmp_file1 = match File::open(&tmp_file_name1) {
			Ok(file) => file,
			Err(error) => {
				eprintln!("{} File open Error: {}", &tmp_file_name1, error);
				exit(1);
			}
		};
		let tmp_file2 = match File::open(&tmp_file_name2) {
			Ok(file) => file,
			Err(error) => {
				eprintln!("{} File open Error: {}", &tmp_file_name2, error);
				exit(1);
			}
		};
		// dst file create
		match File::create(&file_name) {
			Ok(mut dst_file) => {
				let mut reader1 = BufReader::new(tmp_file1);
				let mut reader2 = BufReader::new(tmp_file2);

				let mut byte_cnt: i64 = 0;
				let mut finish_cnt: i32 = 0;
				while finish_cnt != 2 {
					let mut buffer = [0; 1];
					if byte_cnt % 2 == 0 {
						// read 1 byte from file
						match reader1.read(&mut buffer) {
							Ok(0) => { // EOF
								finish_cnt += 1;
								continue;
							},
							Ok(_) => {
								let byte = buffer[0];
								dst_file.write(&[byte]).unwrap();
							}
							Err(error) => {
								eprintln!("tmp file 1read failed: {}", error);
								exit(1);
							}
						}
					} else {
						// read 1 byte from file
						match reader2.read(&mut buffer) {
							Ok(0) => { // EOF
								finish_cnt += 1;
								continue;
							},
							Ok(_) => {
								let byte = buffer[0];
								dst_file.write(&[byte]).unwrap();
							}
							Err(error) => {
								eprintln!("tmp file 1read failed: {}", error);
								exit(1);
							}
						}
					}
					byte_cnt += 1;
				}
				println!("{}{} file merge sucessful!", file_name, file_extension);
				match remove_file(tmp_file_name1) {
					Ok(_) => println!("tmp1 file remove sucess!"),
					Err(_) => {
						eprintln!("tmp1 file remove failed!");
						exit(1);
					}
				};
				match remove_file(tmp_file_name2) {
					Ok(_) => println!("tmp2 file remove sucess!"),
					Err(_) => {
						eprintln!("tmp2 file remove failed!");
						exit(1);
					}
				};
			},
			Err(e) => {
				eprintln!("file creation error: {}", e);
				exit(1);
			}
		}
	} else {
		println!("Invalid argument");
	}
}

fn send_file(file_name: &str, server_name: &str, server_port: &str, part: i32){
	// create client socket
	let mut client_socket = TcpStream::connect(format!("{}:{}", server_name, server_port)).expect("stream failed to connect");
	// prepare command string to send command
	let command_str = "put";
	// write to server
	client_socket.write(command_str.as_bytes()).unwrap();
	// read from server
	let mut command_buffer = vec![0; MSG_SIZE];
	client_socket.read(&mut command_buffer).unwrap();
	let command_res = command_buffer.into_iter().take_while(|&x| x != 0).collect::<Vec<_>>();
	let command_res = String::from_utf8(command_res).expect("invalid utf8 message");
	// server response is not ok
	if command_res != "ok" {
		println!("fail to receive command name");
		exit(1);
	}
	println!("Request to server to put : {}", file_name);

	// file open
	let src_file = match File::open(file_name) {
        Ok(file) => file,
        Err(error) => {
            eprintln!("File open Error: {}", error);
            exit(1);
        }
    };

	// get file extension from fileName
	let file_extension = match Path::new(file_name).extension().and_then(|s| s.to_str()) {
		Some(ext) => ".".to_string() + ext,
		None => "".to_string()
	};
	let file_name = &file_name[0..(file_name.len() - file_extension.len())];
	let file_name = format!("{}-part{}{}", file_name, part + 1, file_extension);
	// write to server
	client_socket.write(file_name.as_bytes()).unwrap();
	// read from server
	let mut filename_buffer = vec![0; MSG_SIZE];
	client_socket.read(&mut filename_buffer).unwrap();
	let filename_res = filename_buffer.into_iter().take_while(|&x| x != 0).collect::<Vec<_>>();
	let filename_res = String::from_utf8(filename_res).expect("invalid utf8 message");
	// server response is not ok
	if filename_res != "ok" {
		println!("fail to receive file name");
		exit(1);
	}
	
	// get file size from File object metadata
	let file_size = match src_file.metadata() {
		Ok(metadata) => metadata.len(),
		Err(_) => 0
	}.to_string();
	client_socket.write(file_size.as_bytes()).unwrap();
	// read from server
	let mut filesize_buffer = vec![0; MSG_SIZE];
	client_socket.read(&mut filesize_buffer).unwrap();
	let filesize_res = filesize_buffer.into_iter().take_while(|&x| x != 0).collect::<Vec<_>>();
	let filesize_res = String::from_utf8(filesize_res).expect("invalid utf8 message");
	// server response is not ok
	if filesize_res != "ok" {
		println!("fail to receive file name");
		exit(1);
	}
	
	let mut reader = BufReader::new(src_file);
    let mut buffer = [0; 1];
	let mut byte_cnt = 0;
    loop {
		// read 1 byte from file
        match reader.read(&mut buffer) {
            Ok(0) => break, // EOF
            Ok(_) => {
                let byte = buffer[0];
                if byte_cnt % 2 == part {
					client_socket.write(&[byte]).unwrap();
				}
            }
            Err(error) => {
                eprintln!("read failed: {}", error);
				exit(1);
            }
        }
		byte_cnt += 1;
    }
	println!("{} send sucessful!", file_name);
}

fn receive_file(file_name: &str, server_name: &str, server_port: &str, part: i32){
	println!("Request to server to get : {}", file_name);

	// create client socket
	let mut client_socket = TcpStream::connect(format!("{}:{}", server_name, server_port)).expect("stream failed to connect");
	// prepare command string to send command
	let command_str = "get";
	client_socket.write(command_str.as_bytes()).unwrap();
	// read from server
	let mut command_buffer = vec![0; MSG_SIZE];
	client_socket.read(&mut command_buffer).unwrap();
	let command_res = command_buffer.into_iter().take_while(|&x| x != 0).collect::<Vec<_>>();
	let command_res = String::from_utf8(command_res).expect("invalid utf8 message");
	// server response is not ok
	if command_res != "ok" {
		println!("fail to receive command name");
		exit(1);
	}

	// get file extension from fileName
	let file_extension = match Path::new(file_name).extension().and_then(|s| s.to_str()) {
		Some(ext) => ".".to_string() + ext,
		None => "".to_string()
	};
	let file_name = &file_name[0..(file_name.len() - file_extension.len())];
	let file_name = format!("{}-part{}{}", file_name, part + 1, file_extension);
	// send file name to server
	client_socket.write(file_name.as_bytes()).unwrap();

	// try to get file size
	let mut filesize_buffer = vec![0; MSG_SIZE];
	client_socket.read(&mut filesize_buffer).unwrap();
	let filesize_res = filesize_buffer.into_iter().take_while(|&x| x != 0).collect::<Vec<_>>();
	let filesize = String::from_utf8(filesize_res).expect("invalid utf8 message");
	// parse file size string to long
	let file_size: u64 = match filesize.parse() {
		Ok(n) => {
			client_socket.write("ok".as_bytes()).unwrap();
			n
		}
		Err(e) => {
            eprintln!("convert error : {}", e);
            exit(1);
        }
	};

	// file create
	match File::create(format!("{}tmp{}", &file_name, file_extension)) {
		Ok(mut file) => {
			let mut received_bytes = 0;
			let mut buffer = vec![0; MSG_SIZE];
			while received_bytes < file_size {
				match client_socket.read(&mut buffer) {
					Ok(n) => {
						if n == 0 {
							break;
						}
						file.write_all(&buffer[..n]).unwrap();
						received_bytes += n as u64;
					},
					Err(e) => {
						println!("Error occurred: {:?}", e);
						continue;
					}
				}
			}
		},
		Err(error) => {
			eprintln!("File open Error: {}", error);
			exit(1);
		}
	};
	println!("{} file store successful!", file_name);

}